File: services/client/src/protocol/protocol.go
package protocol

import (
	"io"
	"strings"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	ACK_SUCCESS_VALUE = 1
	HEADER_SIZE_BYTES = 4

	SHIFT_BYTE_3 = 24
	SHIFT_BYTE_2 = 16
	SHIFT_BYTE_1 = 8
)

// intToBytes converts a uint32 integer to a byte slice of length 4, representing the integer in big-endian order.
func intToBytes(n uint32) []byte {
	return []byte{
		byte(n >> SHIFT_BYTE_3),
		byte(n >> SHIFT_BYTE_2),
		byte(n >> SHIFT_BYTE_1),
		byte(n),
	}
}

// bytesToInt converts a byte slice of length 4 to a uint32 integer, interpreting the bytes in big-endian order.
func bytesToInt(b []byte) uint32 {
	return (uint32(b[0]) << SHIFT_BYTE_3) |
		(uint32(b[1]) << SHIFT_BYTE_2) |
		(uint32(b[2]) << SHIFT_BYTE_1) |
		uint32(b[3])
}

// SendHeader sends a 4-byte header containing the length of the message to the provided writer. It uses the intToBytes function to convert the uint32 value to a byte slice before sending it.
func SendHeader(rw io.Writer, value uint32) error {
	return safe_socket.SendAll(rw, intToBytes(value))
}

// SendStringMessage sends a string message over the provided writer. It first sends a header containing the length of the message, followed by the message itself.
func SendStringMessage(rw io.Writer, msg string) error {
	msgBytes := []byte(msg)
	msgLen := uint32(len(msgBytes))

	packet := make([]byte, HEADER_SIZE_BYTES+msgLen)
	copy(packet[0:HEADER_SIZE_BYTES], intToBytes(msgLen))
	copy(packet[HEADER_SIZE_BYTES:], msgBytes)

	return safe_socket.SendAll(rw, packet)
}

// SendBatch sends a batch of string messages over the provided writer. It concatenates the messages into a single string, separated by newlines, and sends it using the SendStringMessage function.
func SendBatch(rw io.Writer, batch []string) error {
	if len(batch) == 0 {
		return nil
	}
	payload := strings.Join(batch, "\n")
	return SendStringMessage(rw, payload)
}

// RecvStringMessage receives a string message from the provided reader. It first reads a 4-byte header to determine the length of the incoming message, then reads the message itself and returns it as a string.
func RecvStringMessage(r io.Reader) (string, error) {
	header, err := safe_socket.RecvAll(r, HEADER_SIZE_BYTES)
	if err != nil {
		return "", err
	}

	msgLen := bytesToInt(header)
	if msgLen == 0 {
		return "", nil
	}

	payload, err := safe_socket.RecvAll(r, int(msgLen))
	if err != nil {
		return "", err
	}
	return string(payload), nil
}

// RecvACK reads a 4-byte acknowledgment from the provided reader. It checks if the received value matches the expected ACK_SUCCESS_VALUE and returns an error if it doesn't.
func RecvACK(r io.Reader) error {
	header, err := safe_socket.RecvAll(r, HEADER_SIZE_BYTES)
	if err != nil {
		return err
	}
	if bytesToInt(header) != ACK_SUCCESS_VALUE {
		return io.ErrUnexpectedEOF
	}
	return nil
}

File: services/server/src/protocol/protocol.py
import socket
import safe_socket

ACK_CODE = 1
SHIFT_FIRST_BYTE = 24
SHIFT_SECOND_BYTE = 16
SHIFT_THIRD_BYTE = 8

def _int_to_bytes(n: int) -> bytes:
    """
    Convert an integer to a 4-byte big-endian representation.
    """
    return bytes([
        (n >> SHIFT_FIRST_BYTE) & 0xFF,
        (n >> SHIFT_SECOND_BYTE) & 0xFF,
        (n >> SHIFT_THIRD_BYTE) & 0xFF,
        n & 0xFF
    ])

def _bytes_to_int(b: bytes) -> int:
    """
    Convert a 4-byte big-endian representation to an integer.
    """
    return (b[0] << SHIFT_FIRST_BYTE) | (b[1] << SHIFT_SECOND_BYTE) | (b[2] << SHIFT_THIRD_BYTE) | b[3]

def send_header(sock: socket.socket, value: int):
    """
    Send a 4-byte header representing the integer `value` to the socket.
    """
    safe_socket.send_all(sock, _int_to_bytes(value))

def send_string_message(sock: socket.socket, msg: str):
    """
    Send a string message to the socket, prefixed with its length as a 4-byte header.
    """
    msg_bytes = msg.encode('utf-8')
    header = _int_to_bytes(len(msg_bytes))
    safe_socket.send_all(sock, header + msg_bytes)

def recv_header(sock: socket.socket) -> int | None:
    """
    Receive a 4-byte header from the socket and convert it to an integer.
    Returns `None` if the header could not be read.
    """
    header = safe_socket.recv_all(sock, 4)
    if not header or len(header) < 4:
        return None
    return _bytes_to_int(header)

def recv_string_message(sock: socket.socket) -> str | None:
    """
    Receive a string message from the socket, which is prefixed with its length as a 4-byte header.
    Returns `None` if the message could not be read.
    """
    msg_len = recv_header(sock)
    if msg_len is None or msg_len == 0:
        return None
    
    payload = safe_socket.recv_all(sock, msg_len)
    return payload.decode('utf-8')

def recv_batch(sock: socket.socket) -> list[str] | None:
    """
    Receive a batch of string messages from the socket, where the batch is represented as a single string with messages separated by newlines.
    Returns `None` if the batch could not be read.
    """
    payload = recv_string_message(sock)
    if not payload:
        return None
    return [line for line in payload.split('\n') if line]

def send_ack(sock: socket.socket):
    """
    Send an acknowledgment to the socket.
    """
    send_header(sock, ACK_CODE)